doc/building.md

```diff
@@ -877,6 +877,36 @@ make run-test-tier1
 For more details on how to run tests, please see **Testing the JDK**
 ([html](testing.html), [markdown](testing.md)).
 
+## Signing
+
+### macOS
+
+Modern versions of macOS require applications to be signed and notarizied before
+distribution. See Apple's documentation for more background on what this means
+and how it works. To help support this, the JDK build can be configured to
+automatically sign all native binaries, and the JDK bundle, with all the options
+needed for successful notarization, as well as all the entitlements required by
+the JDK. To enable `hardened` signing, use configure parameter
+`--with-macosx-codesign=hardened` and configure the signing identity you wish to
+use with `--with-macosx-codesign-identity=<identity>`. The identity refers to a
+signing identity from Apple that needs to be preinstalled on the build host.
+
+When not signing for distribution with the hardened option, the JDK build will
+still attempt to perform `adhoc` signing, to add the special entitlement
+`com.apple.security.get-task-allow` to each binary. This entitlement is required
+to be able to attach to a process or dump its core. Note that adding this
+entitlement makes the build invalid for notarization, so it is only added when
+signing in `debug` mode. To explicitly enable this kind of adhoc signing, use
+configure parameter `--with-macosx-codesign=debug`. It will be enabled by
+default in most cases.
+
+If no macosx codesign option is specified, the default is to try for `hardened`
+signing if the debug level is `release` and either the default identity or the
+specified identity is valid. If hardened isn't possible, then `debug` signing is
+chosen if it works. If nothing works, the codesign step is skipped. Note that on
+`aarch64`, the Xcode linker will always perform a default `adhoc` signing
+without any entitlements, causing attaching and core dumps not to work.
```

Comment:
I'm getting conflicting messages, both from this text and the PR description, with what is reported in the comments to this PR (and the reason for this PR at all, I assume).

Here and elsewhere it sounds like Xcode will always make `adhoc` signing *without any entitlements* on aarch64. If that is done unconditionally, then this fix cannot possible work on aarch64?

I *think* what you mean is that *if no signing is provided*, on x64, Xlink just skips signing, but on aarch64, it will create an adhoc signing without any entitlements. But if singing is provided, on both x64 and aarch64 that will be used instead.

If my understanding is correct, then I think this message needs updating. In fact, since we now add signing with entitlements, maybe we don't need to talk at all about what happens if we happen to build without them, and can just remove the last sentence?

Reply:
Technically, on aarch64, we overwrite the default linker signing using codesign (using the -f (force) flag). I just wanted to include something that warned of the consequences of not letting the build perform the "debug" mode codesign option.

Reply:
I tried to clarify further.